src/software/thunderscope/robot_communication.py

```diff
@@ -128,9 +147,10 @@ def run_primitive_set(self):
                 primitive_set = self.primitive_buffer.get(
                     block=True, timeout=ROBOT_COMMUNICATIONS_TIMEOUT_S
                 )
-
+                # If we wait on primitive, there's a chance we've initiated shutdown already so we need to re-guard
+                if not self.robots_connected_to_fullsystem:
+                    break
```

Comment:
Does this work when we're just using robot diagnostics? It seems like it will just stop sending primitives if we break out of this loop

Reply:
has same behaviour as current master as of this comment (commit https://github.com/UBC-Thunderbots/Software/commit/c5736a1b32a0c3e5b398dfc7cd1d945070dec08b)